### PR TITLE
Add Performance related settings

### LINKED ISSUE #253: Toggleable options for MangoHud, PRIME and/or GameMode
### Role

i play minecraft

### Suggestion

prime/mangohud toggle

### Benefit

easier to get running for new users

### This suggestion is unique

- [X] I have searched the issue tracker and did not find an issue describing my suggestion, especially not one that has been rejected.

### You may use the editor below to elaborate further.

it would be cool if polymc could detect an optimus enviroment and have a toggle to enable runing with prime/ if in windows running with nvidia/amd gpu over intel intergrated
also a toggle for enable mangohud would be cool as well

### PR DESCRIPTION
**Pull request by Jan200101**
Integrates support for Feral Gamemode, discrete GPU support for Mesa and the proprietary Nvidia driver and MangoHud support

closes #253

### CODE REVIEW

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Scrumplex) [inline launcher/minecraft/launch/DirectJavaLaunch.cpp]: Do we need to touch LD_LIBRARY_PATH? From what I can tell, mangohud is generally installed at `/usr/lib{,64}/libMangoHud{,_dlsym}.so`. If it isn't, the user should deal with this manually (e.g. setting `GAME_LIBRARY_PATH` as an env-var for PolyMC, see JavaUtils::CleanEnviroment)

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Scrumplex) [inline launcher/minecraft/launch/DirectJavaLaunch.cpp]: Move this somewhere into `MinecraftInstance`.

It looks like `MinecraftInstance::createEnvironment()` is actually only intended for the pre/post-launch commands.

I would probably prefer if we just rename `MinecraftInstance::createEnvironment()` to `MinecraftInstance::createHookEnvironment()` and add a new method called `MinecraftInstance::createLaunchEnvironment()`

Maybe this:

```cpp
QProcessEnvironment MinecraftInstance::createLaunchEnvironment()
{
    // prepare the process environment
    QProcessEnvironment env = CleanEnviroment();

    // <insert performance options>

    return env;
}

```

**Review by Scrumplex — COMMENTED** [review]
Looks good overall. Thanks! :D

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: Sure, but you would expect that enabling the MangoHud option would automatically deal with that since you explicitly said you want to use MangoHud.

Simpling throwing libMangoHud into LD_PRELOAD does nothing since `/usr/{LIB}/mangohud` is not on the default search path.

Since writing this I've found out that `ld.so` can parse dynamic string tokens so instead of having that long mess of a lib path string we can simply do `":/usr/local/$LIB/mangohud/:/usr/$LIB/mangohud/` and it will correctly deal with it.

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: > Move this somewhere into `MinecraftInstance`.

Yeah that would be better

> It looks like `MinecraftInstance::createEnvironment()` is actually only intended for the pre/post-launch commands.

Is it? it looks more like a generic environment creation function for the startup process...

> I would probably prefer if we just rename `MinecraftInstance::createEnvironment()` to `MinecraftInstance::createHookEnvironment()` and add a new method called `MinecraftInstance::createLaunchEnvironment()`

I think instead of doing that we should leave the `createEnvironment` function alone and make a second one like this:

```cpp
QProcessEnvironment MinecraftInstance::createLaunchEnvironment()
{
    // prepare the process environment
    QProcessEnvironment env = createEnvironment();

    // <insert performance options>

    return env;
```

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Scrumplex) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: Is this the packaging on Fedora? On Arch mangohud is just thrown into `/usr/lib` (no sub-directory)

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Scrumplex) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: The current `createEnvironment()` also adds env vars (`INST_NAME`, `INST_ID`, `INST_DIR`, ...) which shouldn't probably be passed to Minecraft. They probably don't make a functional difference, but along the lines of #606, we should avoid giving Microsoft more data points to uniquely identify users. On the other hand it might be handy for wrapper scripts. Not sure tbh

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: Yeah that is how its packaged for Fedora, I had been unaware that there are different setups for this.

Looking a bit deeper, both configurations are supported by upstream
https://github.com/flightlessmango/MangoHud/blob/a8a0a245e69fbbca5263d2436fd1c04289375498/src/meson.build#L4-L7

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: I think its not within the scope of this PR to change that behavior, though this definetly should be looked at in the future.

File: launcher/minecraft/launch/DirectJavaLaunch.cpp
Comment (Scrumplex) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: Okay, so `":/usr/local/$LIB/:/usr/local/$LIB/mangohud/:/usr/$LIB/mangohud/` should be best if we really want to `local` to LD_LIBRARY_PATH. (Keep in mind that this also affects java and all natives)

Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: > Okay, so `":/usr/local/$LIB/:/usr/local/$LIB/mangohud/:/usr/$LIB/mangohud/` should be best if we really want to `local` to LD_LIBRARY_PATH. (Keep in mind that this also affects java and all natives)

isn't `/usr/local/$LIB` already in the default search path?

Comment (Scrumplex) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: > isn't `/usr/local/$LIB` already in the default search path?

Nope. Most distributions (userbase-wise) don't include it in LD_LIBRARY_PATH. Notably Ubuntu (Debian?) and Arch

Comment (Jan200101) [reply launcher/minecraft/launch/DirectJavaLaunch.cpp]: looking further into it this isn't the case on Red Hat distros either, I've just been cmake was just doing rpath fuckery which made me believe it was.

File: launcher/minecraft/MinecraftInstance.cpp
Comment (Scrumplex) [inline launcher/minecraft/MinecraftInstance.cpp]:
```suggestion
        // Open Source Drivers
```
nit: trailing spaces

File: launcher/ui/pages/global/MinecraftPage.ui
Comment (Scrumplex) [inline launcher/ui/pages/global/MinecraftPage.ui]: Some tooltips would be nice explaining the options.

My suggestions:

> Enable Feral Interactive's GameMode, to potentially improve gaming performance.

> Enable MangoHud's advanced performance overlay.

> Use the discrete GPU instead of the primary GPU.

File: launcher/ui/pages/global/MinecraftPage.cpp
Comment (Scrumplex) [inline launcher/ui/pages/global/MinecraftPage.cpp]: Maybe hide these instead. Disabled options should only communicate that something is not available *currently*. But these will be permanently disabled on other platforms, so they just shouldn't be there at all.

```suggestion
#if !defined(Q_OS_LINUX)
    ui->enableFeralGamemodeCheck->setVisible(false);
    ui->enableMangoHud->setVisible(false);
    ui->useDiscreteGpuCheck->setVisible(false);
#endif
```

File: launcher/ui/pages/instance/InstanceSettingsPage.cpp
Comment (Scrumplex) [inline launcher/ui/pages/instance/InstanceSettingsPage.cpp]:
```suggestion
#if !defined(Q_OS_LINUX)
    ui->enableFeralGamemodeCheck->setVisible(false);
    ui->enableMangoHud->setVisible(false);
    ui->useDiscreteGpuCheck->setVisible(false);
#endif
```

File: launcher/ui/pages/instance/InstanceSettingsPage.ui
Comment (Scrumplex) [inline launcher/ui/pages/instance/InstanceSettingsPage.ui]: tooltips, see above

File: launcher/minecraft/MinecraftInstance.cpp
Comment (Scrumplex) [inline launcher/minecraft/MinecraftInstance.cpp]: There can be other stuff in `/usr/local/lib` that might cause unexpected issues. If the user wants us to load stuff from there, they should define the necessary env vars themselves.

```suggestion
        auto lib_path = env.value("LD_LIBRARY_PATH", "") +  ":/usr/local/$LIB/mangohud/:/usr/$LIB/mangohud/";
```

File: launcher/ui/pages/instance/InstanceSettingsPage.cpp
Comment (Jan200101) [reply launcher/ui/pages/instance/InstanceSettingsPage.cpp]: for the time being, wouldn't it be better to hide the whole performance group?

File: launcher/ui/pages/global/MinecraftPage.ui
Comment (Jan200101) [reply launcher/ui/pages/global/MinecraftPage.ui]: I went with your suggestions because I cannot come up with anything better.

**Review by Scrumplex — APPROVED** [review]

**Review by flowln — APPROVED** [review]
it would be cool if the options were disabled if the user didn't have the necessary packages installed (or if they don't have a discrete gpu), but maybe it's more trouble than it's worth to implement, so it's fine

thank u!

**Comment by Jan200101** [comment]
> it would be cool if the options were disabled if the user didn't have the necessary packages installed (or if they don't have a discrete gpu), but maybe it's more trouble than it's worth to implement, so it's fine

I am open to implement that if there is a good way to do it.

For gamemode we could simply query the current status and disable it if its an error code.

I don't know what would could be done for the discrete gpu.

**Comment by timoreo22** [comment]
For the discrete gpu, an option would be using something like lspci (or using some of the GPLv2 licensed code that is in it)

lspci has the option -mm for "machine readable output" trying to search for "VGA comppatible controller" should list all gpus, if 2 or more are detected, it should be available

Relevant source code for it [here](https://github.com/pciutils/pciutils/blob/master/lspci.c)

**Comment by Jan200101** [comment]
> For the discrete gpu, an option would be using something like lspci (or using some of the GPLv2 licensed code that is in it)

I don't think shelling out to lspci is the right way to do this.
We could copy the code if really needed but we will have to see

> lspci has the option -mm for "machine readable output" trying to search for "VGA comppatible controller" should list all gpus, if 2 or more are detected, it should be available

This will probably cause false positives on Desktop Computers with a Dedicated GPU and integrated Graphics where the Dedicated GPU is the default.

**Comment by timoreo22** [comment]
Personally when i have my dGPU as main gpu, the iGPU is disabled and replaced with a dummy device (that dosn't count)

**Comment by flowln** [comment]
https://github.com/PolyMC/PolyMC/blob/6cd4375aff98554c56a0138208d869aca0587656/launcher/minecraft/launch/PrintInstanceInfo.cpp#L38

i guess we already have `lspci` stuff lol :p

**Review by txtsd — APPROVED** [review]

**Comment by DioEgizio** [comment]
> For the discrete gpu, an option would be using something like lspci (or using some of the GPLv2 licensed code that is in it)
> 
> lspci has the option -mm for "machine readable output" trying to search for "VGA comppatible controller" should list all gpus, if 2 or more are detected, it should be available
> 
> Relevant source code for it [here](https://github.com/pciutils/pciutils/blob/master/lspci.c)

Gpl2 isn't automatically compatible with the GPL3, and it's incompatible with gpl3 if there's no later version clause, like in this case

**Comment by Scrumplex** [comment]
@Jan200101 You need to sign-off your commit. Otherwise we can't merge it. See https://github.com/PolyMC/PolyMC/pull/707/checks?check_run_id=7208080226

**Comment by Jan200101** [comment]
That is annoying, will do in a second

**Comment by Scrumplex** [comment]
Thanks!

**Pull request merged**
